File: backend/src/search/search.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { Prisma, VerificationStatus } from '@prisma/client';

export interface SearchFilters {
  query?: string;
  specialties?: string[];
  modalities?: string[];
  postcode?: string;
  maxDistance?: number; // km
  minPrice?: number; // pence
  maxPrice?: number; // pence
  availableFor?: 'in-person' | 'remote' | 'both';
  minRating?: number;
  sortBy?: 'rating' | 'distance' | 'price' | 'experience';
  sortOrder?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

export interface PractitionerSearchResult {
  id: string;
  userId: string;
  name: string;
  practiceName?: string;
  bio?: string;
  profilePhoto?: string;
  postcode?: string;
  specialties: string[];
  modalities: string[];
  languagesSpoken: string[];
  yearsOfPractice?: number;
  verifiedBadges: string[];
  sanaIndexScore: number;
  averageRating: number;
  totalReviews: number;
  minPrice?: number; // pence
  maxPrice?: number; // pence
  sessionTypes: Array<{
    id: string;
    name: string;
    duration: number;
    price: number;
    availableFor: string;
  }>;
  distance?: number; // km (if postcode provided)
}

@Injectable()
export class SearchService {
  private readonly logger = new Logger(SearchService.name);

  constructor(private prisma: PrismaService) {}

  /**
   * Search for practitioners with advanced filters
   */
  async searchPractitioners(filters: SearchFilters): Promise<{
    results: PractitionerSearchResult[];
    total: number;
    filters: SearchFilters;
  }> {
    const {
      query,
      specialties,
      modalities,
      postcode,
      maxDistance = 50,
      minPrice,
      maxPrice,
      availableFor,
      minRating,
      sortBy = 'rating',
      sortOrder = 'desc',
      limit = 20,
      offset = 0,
    } = filters;

    // Build Prisma where clause
    const where: Prisma.PractitionerProfileWhereInput = {
      verificationStatus: VerificationStatus.APPROVED,
      user: {
        role: 'PRACTITIONER',
      },
    };

    // Text search (name, bio, specialties, modalities)
    if (query && query.trim() !== '') {
      const searchTerm = query.trim().toLowerCase();
      where.OR = [
        { user: { name: { contains: searchTerm, mode: 'insensitive' } } },
        { practiceName: { contains: searchTerm, mode: 'insensitive' } },
        { bio: { contains: searchTerm, mode: 'insensitive' } },
        { specialties: { has: searchTerm } },
        { modalities: { has: searchTerm } },
      ];
    }

    // Filter by specialties
    if (specialties && specialties.length > 0) {
      where.specialties = { hasSome: specialties };
    }

    // Filter by modalities
    if (modalities && modalities.length > 0) {
      where.modalities = { hasSome: modalities };
    }

    // Filter by postcode (basic implementation - in production use PostGIS or external geocoding API)
    if (postcode) {
      // For now, just filter by exact match or prefix
      // In production, you'd calculate actual distance using lat/lng
      where.postcode = { startsWith: postcode.substring(0, 2) };
    }

    // Fetch practitioners with session types
    const practitioners = await this.prisma.practitionerProfile.findMany({
      where,
      include: {
        user: {
          select: {
            id: true,
            name: true,
            profilePhoto: true,
          },
        },
        sessionTypes: {
          where: { isActive: true },
          select: {
            id: true,
            name: true,
            duration: true,
            price: true,
            availableFor: true,
          },
        },
        reviews: {
          where: { isPublished: true },
          select: {
            rating: true,
          },
        },
      },
    });

    // Apply additional filters and calculate metadata
    let results: PractitionerSearchResult[] = practitioners
      .map((p) => {
        // Calculate average rating
        const ratings = p.reviews.map((r) => r.rating);
        const averageRating =
          ratings.length > 0
            ? ratings.reduce((sum, r) => sum + r, 0) / ratings.length
            : 0;

        // Calculate price range from session types
        const prices = p.sessionTypes.map((st) => st.price);
        const minSessionPrice = prices.length > 0 ? Math.min(...prices) : undefined;
        const maxSessionPrice = prices.length > 0 ? Math.max(...prices) : undefined;

        // Calculate distance (placeholder - in production use actual geocoding)
        let distance: number | undefined;
        if (postcode && p.postcode) {
          distance = this.calculateDistance(postcode, p.postcode);
        }

        return {
          id: p.id,
          userId: p.userId,
          name: p.user.name,
          practiceName: p.practiceName,
          bio: p.bio,
          profilePhoto: p.user.profilePhoto,
          postcode: p.postcode,
          specialties: p.specialties,
          modalities: p.modalities,
          languagesSpoken: p.languagesSpoken,
          yearsOfPractice: p.yearsOfPractice,
          verifiedBadges: p.verifiedBadges,
          sanaIndexScore: p.sanaIndexScore,
          averageRating,
          totalReviews: ratings.length,
          minPrice: minSessionPrice,
          maxPrice: maxSessionPrice,
          sessionTypes: p.sessionTypes,
          distance,
        };
      })
      // Filter by price range
      .filter((p) => {
        if (minPrice !== undefined && p.minPrice !== undefined && p.minPrice < minPrice) {
          return false;
        }
        if (maxPrice !== undefined && p.maxPrice !== undefined && p.maxPrice > maxPrice) {
          return false;
        }
        return true;
      })
      // Filter by availability type
      .filter((p) => {
        if (!availableFor) return true;
        return p.sessionTypes.some((st) => {
          if (availableFor === 'both') return true;
          return st.availableFor === availableFor || st.availableFor === 'both';
        });
      })
      // Filter by minimum rating
      .filter((p) => {
        if (minRating === undefined) return true;
        return p.averageRating >= minRating;
      })
      // Filter by max distance
      .filter((p) => {
        if (!postcode || p.distance === undefined) return true;
        return p.distance <= maxDistance;
      });

    // Sort results
    results = this.sortResults(results, sortBy, sortOrder);

    // Get total before pagination
    const total = results.length;

    // Apply pagination
    results = results.slice(offset, offset + limit);

    this.logger.log(
      `Search completed: ${results.length} results (${total} total) with filters: ${JSON.stringify(filters)}`,
    );

    return {
      results,
      total,
      filters,
    };
  }

  /**
   * Get available filter options (for UI dropdowns)
   */
  async getFilterOptions(): Promise<{
    specialties: string[];
    modalities: string[];
    priceRange: { min: number; max: number };
  }> {
    // Get all unique specialties
    const practitioners = await this.prisma.practitionerProfile.findMany({
      where: { verificationStatus: VerificationStatus.APPROVED },
      select: { specialties: true, modalities: true },
    });

    const specialtiesSet = new Set<string>();
    const modalitiesSet = new Set<string>();

    practitioners.forEach((p) => {
      p.specialties.forEach((s) => specialtiesSet.add(s));
      p.modalities.forEach((m) => modalitiesSet.add(m));
    });

    // Get price range
    const sessionTypes = await this.prisma.sessionType.findMany({
      where: { isActive: true },
      select: { price: true },
    });

    const prices = sessionTypes.map((st) => st.price);
    const minPrice = prices.length > 0 ? Math.min(...prices) : 0;
    const maxPrice = prices.length > 0 ? Math.max(...prices) : 20000; // £200 default max

    return {
      specialties: Array.from(specialtiesSet).sort(),
      modalities: Array.from(modalitiesSet).sort(),
      priceRange: { min: minPrice, max: maxPrice },
    };
  }

  /**
   * Get popular search terms (for autocomplete)
   */
  async getPopularSearchTerms(): Promise<string[]> {
    // In production, this would track actual searches
    // For now, return popular specialties and modalities
    const options = await this.getFilterOptions();
    return [...options.specialties.slice(0, 10), ...options.modalities.slice(0, 10)];
  }

  /**
   * Sort results by specified criteria
   */
  private sortResults(
    results: PractitionerSearchResult[],
    sortBy: string,
    sortOrder: 'asc' | 'desc',
  ): PractitionerSearchResult[] {
    const sorted = [...results];

    sorted.sort((a, b) => {
      let comparison = 0;

      switch (sortBy) {
        case 'rating':
          comparison = a.averageRating - b.averageRating;
          // If ratings are equal, sort by number of reviews
          if (comparison === 0) {
            comparison = a.totalReviews - b.totalReviews;
          }
          break;
        case 'distance':
          if (a.distance !== undefined && b.distance !== undefined) {
            comparison = a.distance - b.distance;
          }
          break;
        case 'price':
          if (a.minPrice !== undefined && b.minPrice !== undefined) {
            comparison = a.minPrice - b.minPrice;
          }
          break;
        case 'experience':
          if (a.yearsOfPractice !== undefined && b.yearsOfPractice !== undefined) {
            comparison = a.yearsOfPractice - b.yearsOfPractice;
          }
          break;
        default:
          comparison = 0;
      }

      return sortOrder === 'asc' ? comparison : -comparison;
    });

    return sorted;
  }

  /**
   * Calculate distance between two postcodes (placeholder)
   * In production, use actual geocoding API or PostGIS
   */
  private calculateDistance(postcode1: string, postcode2: string): number {
    // Placeholder: return random distance for demo
    // In production, use Google Maps API, Postcodes.io, or PostGIS
    const prefix1 = postcode1.substring(0, 2);
    const prefix2 = postcode2.substring(0, 2);

    if (prefix1 === prefix2) {
      return Math.random() * 10; // 0-10 km for same area
    }

    return Math.random() * 50 + 10; // 10-60 km for different areas
  }
}

File: frontend/lib/data/services/search_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../models/search.dart';
import 'api_service.dart';

/// Provider for SearchService
final searchServiceProvider = Provider<SearchService>((ref) {
  final apiService = ref.watch(apiServiceProvider);
  return SearchService(apiService);
});

/// Service for practitioner search and filtering
class SearchService {
  final ApiService _apiService;

  SearchService(this._apiService);

  /// Search for practitioners with filters
  Future<SearchResponse> searchPractitioners(SearchFilters filters) async {
    try {
      // Build query parameters
      final Map<String, dynamic> queryParams = {};

      if (filters.query != null && filters.query!.isNotEmpty) {
        queryParams['query'] = filters.query;
      }
      if (filters.specialties != null && filters.specialties!.isNotEmpty) {
        queryParams['specialties'] = filters.specialties!.join(',');
      }
      if (filters.modalities != null && filters.modalities!.isNotEmpty) {
        queryParams['modalities'] = filters.modalities!.join(',');
      }
      if (filters.postcode != null && filters.postcode!.isNotEmpty) {
        queryParams['postcode'] = filters.postcode;
      }
      queryParams['maxDistance'] = filters.maxDistance;

      if (filters.minPrice != null) {
        queryParams['minPrice'] = filters.minPrice;
      }
      if (filters.maxPrice != null) {
        queryParams['maxPrice'] = filters.maxPrice;
      }
      if (filters.availableFor != null && filters.availableFor!.isNotEmpty) {
        queryParams['availableFor'] = filters.availableFor;
      }
      if (filters.minRating != null) {
        queryParams['minRating'] = filters.minRating;
      }

      queryParams['sortBy'] = filters.sortBy;
      queryParams['sortOrder'] = filters.sortOrder;
      queryParams['limit'] = filters.limit;
      queryParams['offset'] = filters.offset;

      final response = await _apiService.dio.get(
        '/search/practitioners',
        queryParameters: queryParams,
      );

      return SearchResponse.fromJson(response.data);
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'Failed to search practitioners',
        );
      } else {
        throw Exception('Network error: ${e.message}');
      }
    }
  }

  /// Get filter options for the search UI
  Future<FilterOptions> getFilterOptions() async {
    try {
      final response = await _apiService.dio.get('/search/filter-options');
      return FilterOptions.fromJson(response.data);
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'Failed to load filter options',
        );
      } else {
        throw Exception('Network error: ${e.message}');
      }
    }
  }

  /// Get popular search terms for autocomplete
  Future<List<String>> getPopularSearchTerms() async {
    try {
      final response = await _apiService.dio.get('/search/popular-terms');
      final data = PopularTermsResponse.fromJson(response.data);
      return data.terms;
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'Failed to load popular terms',
        );
      } else {
        throw Exception('Network error: ${e.message}');
      }
    }
  }

  /// Format price in pence to pounds string
  static String formatPrice(int pence) {
    return '£${(pence / 100).toStringAsFixed(2)}';
  }

  /// Format distance
  static String formatDistance(double distance) {
    if (distance < 1) {
      return '${(distance * 1000).toStringAsFixed(0)}m';
    }
    return '${distance.toStringAsFixed(1)}km';
  }

  /// Get display text for availability type
  static String getAvailabilityText(String availableFor) {
    switch (availableFor) {
      case 'in-person':
        return 'In-person';
      case 'remote':
        return 'Remote';
      case 'both':
        return 'Both';
      default:
        return availableFor;
    }
  }
}

/// State notifier for search state
class SearchState {
  final SearchResponse? response;
  final FilterOptions? filterOptions;
  final List<String>? popularTerms;
  final bool isLoading;
  final String? error;

  SearchState({
    this.response,
    this.filterOptions,
    this.popularTerms,
    this.isLoading = false,
    this.error,
  });

  SearchState copyWith({
    SearchResponse? response,
    FilterOptions? filterOptions,
    List<String>? popularTerms,
    bool? isLoading,
    String? error,
  }) {
    return SearchState(
      response: response ?? this.response,
      filterOptions: filterOptions ?? this.filterOptions,
      popularTerms: popularTerms ?? this.popularTerms,
      isLoading: isLoading ?? this.isLoading,
      error: error,
    );
  }
}

/// State notifier for managing search
class SearchNotifier extends StateNotifier<SearchState> {
  final SearchService _searchService;

  SearchNotifier(this._searchService) : super(SearchState());

  /// Initialize search (load filter options and popular terms)
  Future<void> initialize() async {
    try {
      state = state.copyWith(isLoading: true, error: null);

      final filterOptions = await _searchService.getFilterOptions();
      final popularTerms = await _searchService.getPopularSearchTerms();

      state = state.copyWith(
        filterOptions: filterOptions,
        popularTerms: popularTerms,
        isLoading: false,
      );
    } catch (e) {
      state = state.copyWith(isLoading: false, error: e.toString());
    }
  }

  /// Search with filters
  Future<void> search(SearchFilters filters) async {
    try {
      state = state.copyWith(isLoading: true, error: null);

      final response = await _searchService.searchPractitioners(filters);

      state = state.copyWith(response: response, isLoading: false);
    } catch (e) {
      state = state.copyWith(isLoading: false, error: e.toString());
    }
  }

  /// Load more results (pagination)
  Future<void> loadMore(SearchFilters filters) async {
    if (state.response == null || state.isLoading) return;

    final currentResults = state.response!.results;
    if (currentResults.length >= state.response!.total) return;

    try {
      state = state.copyWith(isLoading: true, error: null);

      final newFilters = filters.copyWith(
        offset: currentResults.length,
      );

      final response = await _searchService.searchPractitioners(newFilters);

      final updatedResponse = state.response!.copyWith(
        results: [...currentResults, ...response.results],
      );

      state = state.copyWith(response: updatedResponse, isLoading: false);
    } catch (e) {
      state = state.copyWith(isLoading: false, error: e.toString());
    }
  }

  /// Clear search results
  void clearResults() {
    state = state.copyWith(response: null, error: null);
  }
}

/// Provider for search state notifier
final searchNotifierProvider =
    StateNotifierProvider<SearchNotifier, SearchState>((ref) {
  final searchService = ref.watch(searchServiceProvider);
  return SearchNotifier(searchService);
});
